device: Fixed CPP compatiblity for driver instantation

vChavezB opened the pull request:
The macro `DEVICE_DT_INST_DEFINE` is not compatible with C++.
    There are two macros that cause this problem.

    1. `Z_DEVICE_INIT` : Wrong order of initialization in struct `device`.
    2. `Z_DEVICE_HANDLES_DEFINE` : By using the __weak attribute with the
    variable declared with `Z_DEVICE_HANDLES_NAME`, C++ throws the error
    "weak declaration must be public".

    This commit fixes the issue by first rearranging the order of
    initialization of the members in the macro `Z_DEVICE_INIT`.
    Secondly, a macro called `Z_DEVICE_HANDLES_EXTERN` has been added.
    This allows to declares the variable defined by `Z_DEVICE_HANDLES_NAME`
    as `extern` if used in a CPP context.

    Signed-off-by: Victor Chavez <chavez-bermudez@fh-aachen.de>

Fixes #54739
Fixes #54737

Disclaimer: I have recreated the pull-request because it seems I made a mistake while doing a force push and the pull-request was closed https://github.com/zephyrproject-rtos/zephyr/pull/54762, sorry!

Commit:
device: Fixed CPP compatiblity for Z_DEVICE_INIT

The macro `Z_DEVICE_INIT` is not compatible with C++ due to
initialization order of the struct device. This commit fixes
the initialization for compatiblity with C++.

Signed-off-by: Victor Chavez <chavez-bermudez@fh-aachen.de>

MaureenHelm commented:
@stephanosio ping

stephanosio reviewed (changes requested):
> 1. `Z_DEVICE_INIT` : Wrong order of initialization in struct `device`.
> 2. `Z_DEVICE_HANDLES_DEFINE` : By using the __weak attribute with the
variable declared with `Z_DEVICE_HANDLES_NAME`, C++ throws the error
"weak declaration must be public".

Please put these two orthogonal changes into separate commits,

Commit:
device: Fixed CPP compatiblity for Z_DEVICE_HANDLES_DEFINE

By using the __weak attribute with the macro `Z_DEVICE_HANDLES_DEFINE`
C++ throws the error "weak declaration must be public". This change
adds the keyword extern if the macro is included in a C++ source
file.

Signed-off-by: Victor Chavez <chavez-bermudez@fh-aachen.de>

vChavezB commented:
> > 1. `Z_DEVICE_INIT` : Wrong order of initialization in struct `device`.
> > 2. `Z_DEVICE_HANDLES_DEFINE` : By using the __weak attribute with the
> >    variable declared with `Z_DEVICE_HANDLES_NAME`, C++ throws the error
> >    "weak declaration must be public".
> 
> Please put these two orthogonal changes into separate commits,

changed as requested
